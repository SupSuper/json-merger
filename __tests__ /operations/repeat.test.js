const jsonMerger = require("../../dist");
const {testConfig} = require("../__helpers__");

describe("when merging two objects and a source property has a $repeat operation", function () {

    test("it should repeat the value", function () {

        const object1 = {
            "prop": []
        };

        const object2 = {
            "prop": {
                "$repeat": {
                    "from": 1,
                    "until": 5,
                    "value": "this should be visible five times"
                }
            }
        };

        const result = jsonMerger.mergeObjects([object1, object2], testConfig());

        expect(result).toMatchSnapshot();
    });

    test("it should merge the result with the target", function () {

        const object1 = {
            "prop": [
                1,
                2,
                3
            ]
        };

        const object2 = {
            "prop": {
                "$repeat": {
                    "from": 1,
                    "until": 1,
                    "value": "this should be the value of the first item"
                }
            }
        };

        const result = jsonMerger.mergeObjects([object1, object2], testConfig());

        expect(result).toMatchSnapshot();
    });

    test("it should set a $repeat local on the scope with the current index", function () {

        const object1 = {
            "prop": []
        };

        const object2 = {
            "prop": {
                "$repeat": {
                    "from": 1,
                    "until": 5,
                    "value": {
                        "$expression": "'This is item with index ' + $repeat.index"
                    }
                }
            }
        };

        const result = jsonMerger.mergeObjects([object1, object2], testConfig());

        expect(result).toMatchSnapshot();
    });

    test("it should set a $repeat local on the scope with the current value", function () {

        const object1 = {
            "prop": []
        };

        const object2 = {
            "prop": {
                "$repeat": {
                    "from": 1,
                    "until": 5,
                    "value": {
                        "$expression": "'Item with index ' + $repeat.index + ' key ' + $repeat.key + ' and value ' + $repeat.value"
                    }
                }
            }
        };

        const result = jsonMerger.mergeObjects([object1, object2], testConfig());

        expect(result).toMatchSnapshot();
    });

    test("and $repeat.range is set it should use the property to repeat a range", function () {

        const object1 = {
            "prop": []
        };

        const object2 = {
            "prop": {
                "$repeat": {
                    "range": "1, 10-12, 20 30, 40-42 50,60  70,,80,90-92,100-102",
                    "value": {
                        "$expression": "'Item with index ' + $repeat.index + ' key ' + $repeat.key + ' and value ' + $repeat.value"
                    }
                }
            }
        };

        const result = jsonMerger.mergeObjects([object1, object2], testConfig());

        expect(result).toMatchSnapshot();
    });

    test("and $repeat.values is set to an array it should use the array to repeat a range", function () {

        const object1 = {
            "prop": []
        };

        const object2 = {
            "prop": {
                "$repeat": {
                    "values": ["a", "b"],
                    "value": {
                        "$expression": "'Item with index ' + $repeat.index + ' key ' + $repeat.key + ' and value ' + $repeat.value"
                    }
                }
            }
        };

        const result = jsonMerger.mergeObjects([object1, object2], testConfig());

        expect(result).toMatchSnapshot();
    });

    test("and $repeat.values is set to an object it should use the property to repeat a range", function () {

        const object1 = {
            "prop": []
        };

        const object2 = {
            "prop": {
                "$repeat": {
                    "values": {
                        "a": "value-of-a",
                        "b": "value-of-b"
                    },
                    "value": {
                        "$expression": "'Item with index ' + $repeat.index + ' key ' + $repeat.key + ' and value ' + $repeat.value"
                    }
                }
            }
        };

        const result = jsonMerger.mergeObjects([object1, object2], testConfig());

        expect(result).toMatchSnapshot();
    });

    test("it should be possible to nest repeat operations and get both indexes from the scope", function () {

        const object1 = {
            "prop": []
        };

        const object2 = {
            "prop": {
                "$repeat": {
                    "from": 0,
                    "until": 1,
                    "value": {
                        "$repeat": {
                            "from": 0,
                            "until": 1,
                            "value": {
                                "$expression": "'This is item ' + $parent.$repeat.index + '.' + $repeat.index"
                            }
                        }
                    }
                }
            }
        };

        const result = jsonMerger.mergeObjects([object1, object2], testConfig());

        expect(result).toMatchSnapshot();
    });
});
