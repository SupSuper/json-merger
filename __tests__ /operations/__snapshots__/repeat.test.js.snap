// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when merging two objects and a source property has a $repeat operation and $repeat.range is set it should use the property to repeat a range 1`] = `
"{
	\\"prop\\": [
		\\"Item with index 0 key 0 and value 1\\",
		\\"Item with index 1 key 1 and value 10\\",
		\\"Item with index 2 key 2 and value 11\\",
		\\"Item with index 3 key 3 and value 12\\",
		\\"Item with index 4 key 4 and value 20\\",
		\\"Item with index 5 key 5 and value 30\\",
		\\"Item with index 6 key 6 and value 40\\",
		\\"Item with index 7 key 7 and value 41\\",
		\\"Item with index 8 key 8 and value 42\\",
		\\"Item with index 9 key 9 and value 50\\",
		\\"Item with index 10 key 10 and value 60\\",
		\\"Item with index 11 key 11 and value 70\\",
		\\"Item with index 12 key 12 and value 80\\",
		\\"Item with index 13 key 13 and value 90\\",
		\\"Item with index 14 key 14 and value 91\\",
		\\"Item with index 15 key 15 and value 92\\",
		\\"Item with index 16 key 16 and value 100\\",
		\\"Item with index 17 key 17 and value 101\\",
		\\"Item with index 18 key 18 and value 102\\"
	]
}"
`;

exports[`when merging two objects and a source property has a $repeat operation and $repeat.values is set to an array it should use the array to repeat a range 1`] = `
"{
	\\"prop\\": [
		\\"Item with index 0 key 0 and value a\\",
		\\"Item with index 1 key 1 and value b\\"
	]
}"
`;

exports[`when merging two objects and a source property has a $repeat operation and $repeat.values is set to an object it should use the property to repeat a range 1`] = `
"{
	\\"prop\\": [
		\\"Item with index 0 key a and value value-of-a\\",
		\\"Item with index 1 key b and value value-of-b\\"
	]
}"
`;

exports[`when merging two objects and a source property has a $repeat operation it should be possible to nest repeat operations and get both indexes from the scope 1`] = `
"{
	\\"prop\\": [
		[
			\\"This is item 0.0\\",
			\\"This is item 0.1\\"
		],
		[
			\\"This is item 1.0\\",
			\\"This is item 1.1\\"
		]
	]
}"
`;

exports[`when merging two objects and a source property has a $repeat operation it should merge the result with the target 1`] = `
"{
	\\"prop\\": [
		\\"this should be the value of the first item\\",
		2,
		3
	]
}"
`;

exports[`when merging two objects and a source property has a $repeat operation it should repeat the value 1`] = `
"{
	\\"prop\\": [
		\\"this should be visible five times\\",
		\\"this should be visible five times\\",
		\\"this should be visible five times\\",
		\\"this should be visible five times\\",
		\\"this should be visible five times\\"
	]
}"
`;

exports[`when merging two objects and a source property has a $repeat operation it should set a $repeat local on the scope with the current index 1`] = `
"{
	\\"prop\\": [
		\\"This is item with index 0\\",
		\\"This is item with index 1\\",
		\\"This is item with index 2\\",
		\\"This is item with index 3\\",
		\\"This is item with index 4\\"
	]
}"
`;

exports[`when merging two objects and a source property has a $repeat operation it should set a $repeat local on the scope with the current value 1`] = `
"{
	\\"prop\\": [
		\\"Item with index 0 key 0 and value 1\\",
		\\"Item with index 1 key 1 and value 2\\",
		\\"Item with index 2 key 2 and value 3\\",
		\\"Item with index 3 key 3 and value 4\\",
		\\"Item with index 4 key 4 and value 5\\"
	]
}"
`;
